stan/math/rev/fun/singular_values.hpp

```diff
@@ -0,0 +1,58 @@
+#ifndef MATH_SINGULAR_VALUES_H
+#define MATH_SINGULAR_VALUES_H
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/functor/reverse_pass_callback.hpp>
+#include <stan/math/rev/core/arena_matrix.hpp>
+#include <stan/math/rev/fun/value_of.hpp>
+#include <stan/math/rev/fun/typedefs.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/err/check_symmetric.hpp>
+#include <stan/math/prim/err/check_nonzero_size.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
+#include <stan/math/prim/fun/value_of_rec.hpp>
+#include <Eigen/SVD>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the singular values of the specified matrix.
+ *
+ * Equation (4) from Differentiable Programming Tensor Networks
+ * is used for MxN input matrix's adjoint calculation.
+ *
+ * @tparam EigMat type of input matrix
+ * @param m MxN input matrix
+ * @return Singular values of matrix
+ */
+template <typename EigMat,
+    require_eigen_matrix_dynamic_t<EigMat>* = nullptr,
+	  require_vt_var<EigMat>* = nullptr>
+inline auto singular_values(const EigMat& m) {
+  using ret_type = promote_scalar_t<var, Eigen::VectorXd>;
+  check_nonzero_size("singular_values", "m", m);
+
+  const int M = to_arena(m.rows());
+  auto arena_m = to_arena(m);
+  
+  Eigen::MatrixXd m_val = value_of(arena_m);
+  Eigen::JacobiSVD<Eigen::MatrixXd> svd(m_val, Eigen::ComputeFullU | Eigen::ComputeFullV);
+
+  vector_v singular_values(to_arena(svd.singularValues())); // switch to vector_v to enable adjoint 
```

Comment:
I think vector_v is shorthand for `Eigen::Matrix<var, Eigen::Dynamic, 1>` (an Eigen vector of vars)

All basic Eigen types allocate on the heap, not the arena. So what is happening is `to_arena(svd.singularValues())` moves a copy of the singular values to the arena as doubles, and then the `vector_v` constructor converts them to vars and constructs an Eigen vector on the heap to store them in.

The thing you want here is:

```
arena_t<ret_type> singular_values = svd.singularValues();
```

Reply:
Gotcha. I really need to read up on the frequently used typedefs haha.

Reply:
> I really need to read up on the frequently used typedefs haha.

The docs on the `arena_t` stuff haven't been written yet :P

Reply:
That is to say not your fault for doing vector_v -- we gotta catch up on the docs